firmware/main.c

```diff
@@ -280,6 +282,84 @@ static void cmd_gpio(BaseSequentialStream *chp, int argc, char *argv[]) {
 	chprintf(chp, "Bad gpio '%s'\r\n", pOpt);
 }
 
+//some variables for the blink thread
+static THD_WORKING_AREA(blinkThreadArea, 128);
+thread_t *blinkThread;
+
+
+static THD_FUNCTION(blinking, arg){
+	int i = (int) arg;
+	//endless loop of blinking lights
+	palSetPadMode(pinPorts[i].gpio, pinPorts[i].pin, PAL_MODE_OUTPUT_PUSHPULL);
+	while (!chThdShouldTerminateX()){
+		palSetPad(pinPorts[i].gpio, pinPorts[i].pin);
+		chThdSleepSeconds(1);
+		palClearPad(pinPorts[i].gpio, pinPorts[i].pin);
+		chThdSleepSeconds(1);
+	}
+	palSetPadMode(pinPorts[i].gpio, pinPorts[i].pin, PAL_MODE_INPUT);
+	return;
+}
+
+
+static void cmd_blink(BaseSequentialStream *chp, int argc, char *argv[]) {
+	//start or stop a thread for the blinking lights
+	int i;
+	char *pOpt = NULL;
+	char *sOpt = NULL;
+
+			//parse arguments
+	for(i = 0; i < argc; i++) {
+		if(strcmp(argv[i], "-s") == 0) {
+			if(++i >= argc) continue;
+#ifdef SHELL_OPTION_PARSING_DEBUG
+			chprintf(chp, "detected -s with val: %s\r\n", argv[i]);
+#endif
+			sOpt = argv[i];}
+
+
+
+			else if(strcmp(argv[i], "-p") == 0) {
+				if(++i >= argc) continue;
+#ifdef SHELL_OPTION_PARSING_DEBUG
+			chprintf(chp, "detected -p with val: %s\r\n", argv[i]);
+#endif
+			pOpt = argv[i];
+	}}
+	//if s is off then pin can be empty
+	if(!sOpt || ((strcmp(sOpt, "off") == 0) && pOpt) || ((strcmp(sOpt, "on") == 0) && !pOpt)){
+		chprintf(chp, "Usage: blink -s [state] [-p <pin>] \r\n"
+									"\twith state:\r\n"
+									"\t\ton | off\r\n"
+									"\twith pin (only if state on):\r\n"
+									"\t\t");
+									//all pins available can be blinked
+		for(i = 0; i < sizeof(pinPorts)/sizeof(pinPorts[0]); i++)
+			if(pinPorts[i].as_gpio)
+				chprintf(chp, "%s | ", pinPorts[i].pinNrString);
+
+		chprintf(chp, "\n");
+		return;
+}
+
+
+	//cancel the thread if state is set to mode
+	if (strcmp(sOpt, "off") == 0) chThdTerminate(blinkThread);
+
+	else if (strcmp(sOpt, "on") == 0){
+	//set the GPIO as output
+	for(i = 0; i < sizeof(pinPorts)/sizeof(pinPorts[0]); i++) {
+		if((pinPorts[i].as_gpio) && (strcmp(pOpt, pinPorts[i].pinNrString) == 0)) {
+
+
+
+			blinkThread = chThdCreateI(blinkThreadArea, sizeof(blinkThreadArea),
```

Comment:
what if blinkThread already is running? are you creating a second thread that overwrites the other thread? (including its stack and working area?)
that must be avoided.
you should initialize blinkThread to NULL. then when either creating or destroying the thread, make sure it is NULL (or != NULL) and set to the thread (or NULL) afterwards.
[in case of terminate, you need to wait for the termination before you can set to NULL, as you only requested the termination and the thread still needs to leave its main function.]

Reply:
this still has some issues. it is important that the firmware is always in a consistent state.
and that no user input or timing-dependent interaction can lead to an inconsistent state.
if i look at the blinker thread and the cmd_blink function i still see ways to break this.
e.g. blinkThread is not initialized to NULL.
also in the terminate-part of the `if (strcmp(sOpt, "off") == 0 || blinkThread != NULL) {` you use a sleep(3), then blinkThread=NULL. what if the thread needs longer than three seconds to terminate? this is very unlikely here but problematic design in general. you need to wait for proper thread termination. use the corresponding functions provided by the RTOS, e.g. chThdTerminatedX().
also that if-clause should print an error if a thread was running and the user requested another blinker.

please also rename the blinker thread from 'blinking' to 'blinkerThread' or something more sensible.
also I don't think "include <pthread.h>" is sensible here. we're using chibios, and its corresponding header (if any) should be included. i don't think we really need any besides <ch.h>